content/client-lib-development-guide/features.textile

```diff
@@ -1464,6 +1486,110 @@ enum StatsIntervalGranularity:
 
+class DeviceDetails:
+  + getDeviceInfo() => io DeviceDetails
+  id: String
+  platform: DeviceType
+  type: DeviceType
+  clientId: String?
+  metadata: Dict<String, String>
+  updateTokenId: String
+  resetId(String)
+  resetUpdateToken() => io
+  push: DevicePushDetails
+
+class DevicePushDetails:
+  transportType: DevicePushTransportType
+  transportId: String
+  transportToken: String?
+  state: .Active | .Failing | .Failed
+  errorReason: ErrorInfo?
+  metadata: Dict<String, String>
+
+class Push:
+  publish(clientId: String, JsonObject) => io
+  publish(deviceId: String, JsonObject) => io
+  publish(transportType: String, transportId: String, transportTokenOrKey?: String) => io
+  admin: PushAdmin
+
+  // Only on platforms that support receiving notifications:
+
+  activate(DeviceDetails) => io ErrorInfo? // RSH2c
+  activateExternally(
+    DeviceDetails,
+    registerDevice: (ErrorInfo?, DeviceDetails?) -> io String?,
+    renewRegistration: (ErrorInfo?, DeviceDetails?) -> io String?
+  )
+  deactivate(deviceId: String) => io ErrorInfo?
+  deactivateExternally
+    deviceId: String,
+    deactivate: (ErrorInfo?, deviceId: String) -> io
+  )
+
+class PushAdmin:
+  deviceRegistrations: PushDeviceRegistrations
+  channelSubscriptions: PushChannelSubscriptions
+
+class JsonObject:
+  // Platform-dependant, typically a Dict-like object
+
+class PushDeviceRegistrations:
+  save(PushDeviceRegistration) => io
+  find(type: PushDeviceType, transportId: String) => io PushDeviceRegistration?
+  get(transportIds: [(PushDeviceType, String)], state?: .Active | .Failed) -> io PaginatedResult<PushDeviceRegistration>device
+  get(clientId: String, state?: .Active | .Failed) -> io PaginatedResult<PushDeviceRegistration>
+  remove(type: PushDeviceType, transportId: String) -> io
+  remove(clientId: String) => io
+  remove(PushDeviceRegistration) => io
+
+class PushChannelSubscriptions:
+  save([PushChannelSubscription]) => io
+  save(PushChannelSubscription) => io
+  listChannels() => io PaginatedResult<String>
+  get(channelName: String) => io PaginatedResult<PushChannelSubscription>
+  get(type: PushDeviceType, transportId: String, channelName: String?) => io PaginatedResult<PushChannelSubscription>
+  get(clientId: String, channelName: String?) => io PaginatedResult<PushChannelSubscription>
+  remove([PushChannelSubscription]) => io
+  remove(PushChannelSubscription) => io
+
+class PushDeviceRegistration:
+  embeds DeviceDetails // PDR1
+  deviceId: String
+  devicePlatform: DevicePlatform
+  deviceType: PushDeviceType
+  transportType: DevicePushTransportType
+  transportId: String,
+  transportTokenOrKey: String?,
+  metadata: Dict<String, String>?
+  clientId: String?, // PDR2
+  state: .Active | .Failed // PDR3
+  errorReason: ErrorInfo? // PDR4
+
+enum DevicePushTransportType:
+  "fcm" // PTT1
+  "gcm" // PTT1
+  "apns" // PTT1
+  "web" // PTT1
+
+enum DevicePlatform:
+  "android" // PPT1
+  "ios" // PPT1
+  "browser" // PPT1
+
+enum DeviceType:
```

Comment:
Type is quite broad.  Should this perhaps be `DeviceFormFactor` instead?  @paddybyers WDYT?